node/make.js

```diff
@@ -59,7 +52,152 @@ target.loc = function() {
+            break;
+    }
+}
+
+var downloadFile = function (url) {
+    // validate parameters
+    if (!url) {
+        throw new Error('Parameter "url" must be set.');
+    }
+
+    // short-circuit if already downloaded
+    var scrubbedUrl = url.replace(/[/\:?]/g, '_');
+    var targetPath = path.join(downloadPath, 'file', scrubbedUrl);
+    var marker = targetPath + '.completed';
+    if (test('-f', marker)) {
+        console.log('Download file already cached: ' + url);
+    }
+    else {
+        console.log('Downloading file: ' + url);
+
+        // delete any previous partial attempt
+        if (test('-f', targetPath)) {
+            rm('-f', targetPath);
+        }
+
+        // download the file
+        mkdir('-p', path.join(downloadPath, 'file'));
+        var result = syncRequest('GET', url);
+        fs.writeFileSync(targetPath, result.getBody());
+
+        // write the completed marker
+        fs.writeFileSync(marker, '');
+    }
+
+    return targetPath;
+}
+
+var downloadArchive = function (url) {
+    // validate parameters
+    if (!url) {
+        throw new Error('Parameter "url" must be set.');
+    }
+
+    if (!url.match(/\.tar\.gz$/)) {
+        throw new Error('Expected .tar.gz');
+    }
+
+    // short-circuit if already downloaded and extracted
+    var scrubbedUrl = url.replace(/[/\:?]/g, '_');
+    var targetPath = path.join(downloadPath, 'archive', scrubbedUrl);
+    var marker = targetPath + '.completed';
+    if (test('-f', marker)) {
+        console.log('Download archive already cached: ' + url);
+    }
+    else {
+        // download the archive
+        var archivePath = downloadFile(url);
+        console.log('Extracting archive: ' + url);
+
+        // delete any previous temporary tar file
+        var tarPath = targetPath + '.temptar';
+        if (test('-f', tarPath)) {
+            rm('-f', tarPath);
+        }
+
+        // un-gzip
+        mkdir('-p', path.dirname(tarPath));
+        var content = fs.readFileSync(archivePath);
+        content = zlib.gunzipSync(content);
+        fs.writeFileSync(tarPath, content);
+
+        // delete any previously attempted extraction directory
+        if (test('-d', targetPath)) {
+            rm('-rf', targetPath);
+        }
+
+        // un-tar
+        mkdir('-p', targetPath);
+        fs.createReadStream(tarPath).pipe(tar.extract(targetPath));
+
+        // delete the temporary tar file
+        rm('-f', tarPath);
+
+        // write the completed marker
+        fs.writeFileSync(marker, '');
+    }
+
+    return targetPath;
+}
+
+var addPath = function (directory) {
+    var separator;
+    if (os.platform() == 'win32') {
+        separator = ';';
+    }
+    else {
+        separator = ':';
+    }
+
+    var existing = process.env['PATH'];
+    if (existing) {
+        process.env['PATH'] = directory + separator + existing;
+    }
+    else {
+        process.env['PATH'] = directory;
```

Comment:
Trailing or leading ":" on nix means to add the current directory to the PATH (wherever current directory is at the time a cmd is invoked). It is also viewed as a potential security vulnerability - so I took precaution to avoid adding a trailing ":" to the PATH.